satellite/audit/reservoirchore.go

```diff
@@ -0,0 +1,88 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package audit
+
+import (
+	"context"
+	"math/rand"
+	"time"
+
+	"go.uber.org/zap"
+
+	"storj.io/storj/internal/sync2"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/satellite/metainfo"
+)
+
+// ReservoirChore populates reservoirs and the audit queue.
+type ReservoirChore struct {
+	log    *zap.Logger
+	config Config
+	rand   *rand.Rand
+
+	service *Service2
+
+	MetainfoLoop *metainfo.Loop
+	Loop         sync2.Cycle
+	Interval     time.Duration
+}
+
+// NewReservoirChore instantiates ReservoirChore.
+func NewReservoirChore(log *zap.Logger, service *Service2, metaLoop *metainfo.Loop, config Config) *ReservoirChore {
+	return &ReservoirChore{
+		log:    log,
+		config: config,
+		rand:   rand.New(rand.NewSource(time.Now().Unix())),
+
+		service: service,
+
+		MetainfoLoop: metaLoop,
+		Loop:         *sync2.NewCycle(config.QueueInterval),
+		Interval:     config.ChoreInterval,
+	}
+}
+
+func (chore *ReservoirChore) Run(ctx context.Context) error {
+	return chore.Loop.Run(ctx, func(ctx context.Context) (err error) {
+		defer mon.Task()(&ctx)(&err)
+
+		select {
+		case <-chore.service.queue.closed:
```

Comment:
ReservoirChore shouldn't have to care about queue being closed. Queue is a persistence concern so it shouldn't be closed until all services/chores and peer has stopped.

See how repair handles it.

Reply:
I removed the concept of closing from the queue, and the channel is now on the service struct. I do not think the check is necessary since the peer closes the service before the chore, but it's probably safer to have it I think.